From: Michael Schutte <michi@debian.org>
Subject: Various fixes to the openvt utility
Forwarded: no
Bug-Debian: http://bugs.debian.org/190387
Bug-Debian: http://bugs.debian.org/630108
Bug-Debian: http://bugs.debian.org/642324

Three unrelated fixes:

 * Propagate child exit codes in “openvt -w”.

 * consfd stores the file descriptor referring to the console.  In some
   cases, this will be one of the stdio fds, which are closed in the
   child process.  Avoid closing it again.

 * When showing the help output, only exit with a zero status if the
   “-h” option was given.

---
 src/openvt.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

Index: devel/src/openvt.c
===================================================================
--- devel.orig/src/openvt.c	2011-09-23 21:32:28.000000000 +0200
+++ devel/src/openvt.c	2011-09-23 21:34:09.000000000 +0200
@@ -260,10 +260,12 @@
 			case 'V':
 				print_version_and_exit();
 				break;
-			default:
 			case 'h':
 				print_help(EXIT_SUCCESS);
 				break;
+			default:
+				print_help(EXIT_FAILURE);
+				break;
 		}
 	}
 
@@ -385,7 +387,8 @@
 		close(0);
 		close(1);
 		close(2);
-		close(consfd);
+		if (consfd > 2)
+			close(consfd);
 
 		if ((dup2(fd, 0) == -1) || (dup2(fd, 1) == -1) || (dup2(fd, 2) == -1))
 			openvt_fatal(1, errno, "dup");
@@ -408,7 +411,6 @@
 	if (do_wait) {
 		int retval = 0; /* actual value returned form process */
 
-		wait(NULL);
 		waitpid(pid, &retval, 0);
 
 		if (show) {	/* Switch back... */
@@ -425,7 +427,10 @@
 
 		/* if all our stuff went right, we want to return the exit code of the command we ran
 		   super vital for scripting loops etc */
-		return(retval);
+		if (WIFEXITED(retval))
+			return WEXITSTATUS(retval);
+		else
+			return 127;
 	}
 
 	return EXIT_SUCCESS;
